drivers/sgp30/sgp30.c

```diff
@@ -0,0 +1,353 @@
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sgp30
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the sensors
+ *
+ * @author      Francisco Molina <francois-xavier.molina@inria.fr>
+ *
+ * @}
+ */
+#include <string.h>
+
+#include "sgp30.h"
+#include "sgp30_constants.h"
+#include "sgp30_params.h"
+
+#include "checksum/crc8.h"
+#include "ztimer.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+/**
+ * @name      Address pointer values for all SGP30 I2C commands
+ * @{
+ */
+typedef enum {
+    SGP30_CMD_INIT_AIR_QUALITY          = 0x2003,   /**< Start measurement mode */
+    SGP30_CMD_MEASURE_AIR_QUALITY       = 0x2008,   /**< Measure air quality values */
+    SGP30_CMD_GET_BASELINE              = 0x2015,   /**< Get baseline values */
+    SGP30_CMD_SET_BASELINE              = 0x201e,   /**< Set baseline values CMD */
+    SGP30_CMD_SET_HUMIDITY              = 0x2061,   /**< Set absolute humidity value
+                                                         for compensation */
+    SGP30_CMD_MEASURE_TEST              = 0x2032,   /**< Perform on-chip self test */
+    SGP30_CMD_GET_FEATURE_SET_VERSION   = 0x202f,   /**< Get feature set version */
+    SGP30_CMD_MEASURE_RAW_SIGNALS       = 0x2050,   /**< Read raw H2 and Ethanol signals */
+    SGP30_CMD_READ_SERIAL               = 0x3682,   /**< Read serial number */
+    SGP30_CMD_SOFT_RESET                = 0x0006,   /**< Perform General Call reset */
+} sgp30_cmd_t;
+/** @} */
+
+/**
+ * @name      Max delays for SGP30 I2C commands completion
+ * @{
+ */
+#define SGP30_DELAY_INIT_AIR_QUALITY        (10 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_AIR_QUALITY     (12 * MS_PER_SEC)
+#define SGP30_DELAY_GET_BASELINE            (10 * MS_PER_SEC)
+#define SGP30_DELAY_SET_BASELINE            (10 * MS_PER_SEC)
+#define SGP30_DELAY_SET_HUMIDITY            (10 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_TEST            (220 * MS_PER_SEC)
+#define SGP30_DELAY_GET_FEATURE_SET_VERSION (2 * MS_PER_SEC)
+#define SGP30_DELAY_MEASURE_RAW_SIGNALS     (25 * MS_PER_SEC)
+#define SGP30_DELAY_READ_SERIAL             (500)
+#define SGP30_DELAY_SOFT_RESET              (60 * US_PER_MS)
+/** @} */
+
+/* Device power up time */
+#define SGP30_POWER_UP_TIME         (60 * US_PER_MS)
+/* Polynomial used in crc check */
+#define SGP30_CRC8_POLYNOMIAL       0x31
+/* Seed used in crc check */
+#define SGP30_CRC8_SEED             0xFF
+
+static inline uint8_t _crc8(const void *buf, size_t len)
+{
+    return crc8(buf, len, SGP30_CRC8_POLYNOMIAL, SGP30_CRC8_SEED);
+}
+
+static int _rx_tx_data(sgp30_t *dev, uint16_t cmd, uint8_t *data,
+                       size_t len, uint32_t delay, bool read)
+{
+    int res = 0;
+
+    if (i2c_acquire(dev->params.i2c_dev) != 0) {
+        DEBUG("[sgp30]: could not acquire I2C bus %d\n", dev->params.i2c_dev);
+        return -1;
+    }
+
+    uint8_t frame_cmd[sizeof(cmd) + len];
```

Comment:
(style, feel free to ignore)
```suggestion
    uint8_t frame_cmd[sizeof(cmd) + len] = { cmd >> 8, cmd & 0xFF };
```